<!DOCTYPE html>
<html lang="en-us">
<head>

  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="generator" content="Source Themes Academic 3.3.0">
  <meta name="generator" content="Hugo 0.54.0" />
  <meta name="author" content="Krishnaswamy Lab">

  
  
  
  
    
  
  <meta name="description" content="Once you&rsquo;ve inspected the principle components of your dataset, it&rsquo;s time to start visualizing your data using PHATE. We&rsquo;re going to demonstrate PHATE analysis on a few datasets. We will show:
 How PHATE works
 Running PHATE on several datasets
 How to interpret a PHATE plot
 Clustering using the diffusion potential How to pick parameters for PHATE
 How to troubleshoot common issues with PHATE plots">

  
  <link rel="alternate" hreflang="en-us" href="https://dburkhardt.github.io/tutorial/visualizing_phate/">

  


  

  

  

  

  

  

  
  
  
  <meta name="theme-color" content="#2962ff">
  

  
  
  
  
    
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/4.1.3/css/bootstrap.min.css" integrity="sha256-eSi1q2PG6J7g7ib17yAaWMcrr5GrtohYChqibrV7PBE=" crossorigin="anonymous">
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/academicons/1.8.6/css/academicons.min.css" integrity="sha256-uFVgMKfistnJAfoCUQigIl+JfUaP47GrRKjf6CTPVmw=" crossorigin="anonymous">
    <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.6.0/css/all.css" integrity="sha384-aOkxzJ5uQz7WBObEZcHvV5JvRW3TUc2rNPA7pe3AwnsUohiw1Vj2Rgx2KSOkF5+h" crossorigin="anonymous">
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.2.5/jquery.fancybox.min.css" integrity="sha256-ygkqlh3CYSUri3LhQxzdcm0n1EQvH2Y+U5S2idbLtxs=" crossorigin="anonymous">

    
    
    
      
    
    
      
      
        
          <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css" crossorigin="anonymous">
        
      
    

    

    

  

  
  
  <link rel="stylesheet" href="//fonts.googleapis.com/css?family=Montserrat:400,700|Roboto:400,400italic,700|Roboto+Mono">
  

  <link rel="stylesheet" href="/styles.css">
  
  <link rel="stylesheet" href="/css/custom.css">
  

  
  
  

  
  <link rel="alternate" href="https://dburkhardt.github.io/index.xml" type="application/rss+xml" title="Krishnaswamy Lab">
  <link rel="feed" href="https://dburkhardt.github.io/index.xml" type="application/rss+xml" title="Krishnaswamy Lab">
  

  <link rel="manifest" href="/site.webmanifest">
  <link rel="icon" type="image/png" href="/img/icon.png">
  <link rel="apple-touch-icon" type="image/png" href="/img/icon-192.png">

  <link rel="canonical" href="https://dburkhardt.github.io/tutorial/visualizing_phate/">

  
  
  
  
    
    
  
  <meta property="twitter:card" content="summary">
  
  <meta property="og:site_name" content="Krishnaswamy Lab">
  <meta property="og:url" content="https://dburkhardt.github.io/tutorial/visualizing_phate/">
  <meta property="og:title" content="3. Visualizing data using PHATE | Krishnaswamy Lab">
  <meta property="og:description" content="Once you&rsquo;ve inspected the principle components of your dataset, it&rsquo;s time to start visualizing your data using PHATE. We&rsquo;re going to demonstrate PHATE analysis on a few datasets. We will show:
 How PHATE works
 Running PHATE on several datasets
 How to interpret a PHATE plot
 Clustering using the diffusion potential How to pick parameters for PHATE
 How to troubleshoot common issues with PHATE plots"><meta property="og:image" content="https://dburkhardt.github.io/img/portrait.jpg">
  <meta property="og:locale" content="en-us">
  
  <meta property="article:published_time" content="2018-09-09T00:00:00-04:00">
  
  <meta property="article:modified_time" content="2019-01-01T00:00:00-05:00">
  

  

  

  <title>3. Visualizing data using PHATE | Krishnaswamy Lab</title>

</head>
<body id="top" data-spy="scroll" data-target="#TableOfContents" data-offset="71" >
  <aside class="search-results" id="search">
  <div class="container">
    <section class="search-header">

      <div class="row no-gutters justify-content-between mb-3">
        <div class="col-6">
          <h1>Search</h1>
        </div>
        <div class="col-6 col-search-close">
          <a class="js-search" href="#"><i class="fas fa-times-circle text-muted" aria-hidden="true"></i></a>
        </div>
      </div>

      <div id="search-box">
        
        <input name="q" id="search-query" placeholder="Search..." autocapitalize="off"
        autocomplete="off" autocorrect="off" role="textbox" spellcheck="false" type="search">
        
      </div>

    </section>
    <section class="section-search-results">

      <div id="search-hits">
        
      </div>

    </section>
  </div>
</aside>


<nav class="navbar navbar-light fixed-top navbar-expand-lg py-0" id="navbar-main">
  <div class="container">

    
      <a class="navbar-brand" href="/">Krishnaswamy Lab</a>
      
      <button type="button" class="navbar-toggler" data-toggle="collapse"
              data-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
        <span><i class="fas fa-bars"></i></span>
      </button>
      

    
    <div class="collapse navbar-collapse" id="navbar">

      
      
      <ul class="navbar-nav ml-auto">
        

        

        
        
        
          
        

        <li class="nav-item">
          <a class="nav-link" href="/#about">
            
            <span>Home</span>
            
          </a>
        </li>

        
        

        

        
        
        
          
        

        <li class="nav-item">
          <a class="nav-link" href="/post/">
            
            <span>Posts</span>
            
          </a>
        </li>

        
        

        

        
        
        
          
        

        <li class="nav-item">
          <a class="nav-link" href="/tutorial/">
            
            <span>Tutorials</span>
            
          </a>
        </li>

        
        

      

        

        
        <li class="nav-item">
          <a class="nav-link js-search" href="#"><i class="fas fa-search" aria-hidden="true"></i></a>
        </li>
        

        

        
        <li class="nav-item">
          <a class="nav-link js-dark-toggle" href="#"><i class="fas fa-moon" aria-hidden="true"></i></a>
        </li>
        

      </ul>

    </div>
  </div>
</nav>



<div class="container-fluid docs">
  <div class="row flex-xl-nowrap">
    <div class="col-12 col-md-3 col-xl-2 docs-sidebar">
      




<form class="docs-search d-flex align-items-center">
  <button class="btn docs-toggle d-md-none p-0 mr-3" type="button" data-toggle="collapse" data-target="#docs-nav" aria-controls="docs-nav" aria-expanded="false" aria-label="Toggle section navigation">
    <span><i class="fas fa-bars"></i></span>
  </button>

  
  <input name="q" type="search" class="form-control" id="search-query" placeholder="Search..." autocomplete="off">
  
</form>

<nav class="collapse docs-links" id="docs-nav">
  
  
  <div class="docs-toc-item">
    <a class="docs-toc-link" href="/tutorial/">Overview</a>

  </div>
  
  <div class="docs-toc-item">
    <a class="docs-toc-link" href="/tutorial/what_is_single_cell/">How to single cell</a>
    <ul class="nav docs-sidenav">
      
      <li >
        <a href="/tutorial/what_is_single_cell/">1. What is single cell?</a>
      </li>
      
      <li >
        <a href="/tutorial/preprocessing/">2. Preprocessing</a>
      </li>
      
      <li >
        <a href="/tutorial/visualizing_pca/">3. Visualizing PCA dimensions</a>
      </li>
      
      <li class="active">
        <a href="/tutorial/visualizing_phate/">4. Visualizing data using PHATE</a>
      </li>
      
    </ul>
    

  </div>
  
  
</nav>

    </div>

    
    <div class="d-none d-xl-block col-xl-2 docs-toc">
      
      <p class="docs-toc-title">On this page</p>
      

      <nav id="TableOfContents">
<ul>
<li>
<ul>
<li><a href="#3-0-what-is-phate-and-why-should-you-use-it">3.0 - What is PHATE and why should you use it?</a>
<ul>
<li><a href="#installing-phate">Installing PHATE</a></li>
</ul></li>
<li><a href="#3-1-phate-parameters">3.1 PHATE parameters</a></li>
<li><a href="#3-2-wait-a-second-what-s-a-kernel">3.2 <em>Wait a second, what&rsquo;s a kernel?</em></a></li>
<li><a href="#3-3-instantiating-a-phate-estimator">3.3  Instantiating a PHATE estimator</a></li>
<li><a href="#3-4-plotting-phate-with-a-scatter-plot">3.4 Plotting PHATE with a scatter plot</a></li>
<li><a href="#3-5-plotting-phate-as-a-kde-plot">3.5 Plotting PHATE as a KDE plot</a></li>
<li><a href="#3-6-clustering-data-using-phate">3.6 Clustering data using PHATE</a></li>
</ul></li>
</ul>
</nav>

      <ul class="nav toc-top">
        <li><a href="#">Back to top</a></li>
      </ul>

      
    </div>
    

    <main class="col-12 col-md-9 col-xl-8 py-md-3 pl-md-5 docs-content" role="main">

      <article class="article" itemscope itemtype="http://schema.org/Article">

        <div class="docs-article-container">
          <h1 itemprop="name">3. Visualizing data using PHATE</h1>

          <div class="article-style" itemprop="articleBody">
            

<p>Once you&rsquo;ve inspected the principle components of your dataset, it&rsquo;s time to start visualizing your data using PHATE. We&rsquo;re going to demonstrate PHATE analysis on a few datasets. We will show:</p>

<ul>
<li>How PHATE works<br /></li>
<li>Running PHATE on several datasets<br /></li>
<li>How to interpret a PHATE plot<br /></li>
<li>Clustering using the diffusion potential</li>
<li>How to pick parameters for PHATE<br /></li>
<li>How to troubleshoot common issues with PHATE plots<br /></li>
</ul>

<h2 id="3-0-what-is-phate-and-why-should-you-use-it">3.0 - What is PHATE and why should you use it?</h2>

<p>PHATE is a dimensionaltiy reduction developed by the Krishnaswamy lab for visualizing high-dimensional data. We use PHATE for <em>every</em> dataset the comes through the lab: scRNA-seq, CyTOF, gut microbiome profiles, simulated data, etc. PHATE was designed to handle noisy, non-linear relationships between data points. PHATE produces a low-dimensional representation that preserves both local and global structure in a dataset so that you can make generate hypotheses from the plot about the relationships between cells present in a dataset. Although PHATE has utility for analysis of many data modalities, we will focus on the application of PHATE for scRNA-seq analysis.</p>

<p>PHATE is inspired by diffusion maps <a href="https://doi.org/10.1016/j.acha.2006.04.006" target="_blank">(Coifman et al. (2008))</a>, but include several key innovations that make it possible to generate a two or three dimensional visualization that preserves continuous relationships between cells where they exist. For a full explanation of the PHATE algorithm, please consult <a href="https://doi.org/10.1101/120378" target="_blank">the PHATE manuscript</a>.</p>

<h3 id="installing-phate">Installing PHATE</h3>

<p>PHATE is available on PyPI, and can be installed by running the following command in a terminal:</p>

<pre><code class="language-bash">$ pip install --user phate
</code></pre>

<p>PHATE is also available in <code>MATLAB</code> and <code>R</code>, but we&rsquo;re going to focus on the Python implementation for this tutorial.</p>

<h2 id="3-1-phate-parameters">3.1 PHATE parameters</h2>

<p>As we mentioned in the previous section, PHATE is a subclass of the <code>sklearn.BaseEstimator</code>, and it&rsquo;s API matches that of the <code>PCA</code> operator. To use PHATE, you must first instantiate a <code>phate.PHATE</code> object, and then use <code>fit_transform</code> to build a graph to your data and reduce the dimensionality of the data for visualization.</p>

<p>PHATE has three key parameters:</p>

<ul>
<li><code>n_components</code> - sets the number of dimensions to which PHATE will reduce the input data<br /></li>
<li><code>knn</code> - sets the number of Nearest Neighbors to use for calculating the kernel bandwidth<br /></li>
<li><code>decay</code> - sets the rate of decay of the kernel tails<br /></li>
</ul>

<p>I&rsquo;ll introduce other parameters thoughout the tutorial, but you should also check out the full documentation on readthedocs: <a href="https://phate.readthedocs.io/en/stable/" target="_blank">https://phate.readthedocs.io/en/stable/</a>.</p>

<h2 id="3-2-wait-a-second-what-s-a-kernel">3.2 <em>Wait a second, what&rsquo;s a kernel?</em></h2>

<p>Now if you&rsquo;ve never studied graph theory or discrete mathematics, these last two parameters might be confusing. To understand them, consider a very simple graph, the k-Nearest Neighbor (kNN) graph. Here, each cell is a node in the graph, and edges exist between a cell and it&rsquo;s <code>k</code> nearest neighbors. You can also think about this as a graph where all cells are connected, but the connections between non-neighboring cells have a <em>strength</em> or <em>weight</em> of 0, while the edges between neighboring cells have a weight of 1.</p>

<p>The kNN graph offers a very powerful representation for single cell data, but it also has some drawbacks. For example, consider the following graphs:</p>




<figure class="img-sm">

<img src="/img/how_to_single_cell/PHATE_graph_explanation.png" />


</figure>

<p>On the left, we have a kNN graph where <em>k</em>=4. Notice that all the blue cells, regardless of their proximity to the red cell, have edges of equal weight. Also notice that the green cell in the lower right hand corner has no connection to the red cell, despite being only trivially farther away than the next closest cell. These two properties of kNN graphs, a harsh cutoff for edges and uniform edge weights, means that choice of proper <code>k</code> is critical for any method using this graph.</p>

<p>To overcome these limitations, we use a variation of the <a href="https://en.wikipedia.org/wiki/Radial_basis_function_kernel" target="_blank">radial basis kernel</a> graph on the right. This graph connects cells with edge weights proportionally to their distance to the red cell. You can think of this kernel function as a &ldquo;soft&rdquo; kNN, where the weights vary smoothly between 0 and 1. Note: in practice, all cells are connected at one stage during graph instantiation, but then edges that are below a cutoff, such as <code>1e-4</code> are set to 0.</p>

<p>Now that you understand this distinction, you can think of the <code>knn</code> parameter as setting a baseline for finding close cells, and <code>decay</code> as the &ldquo;softness&rdquo; the edge weighting.</p>

<h2 id="3-3-instantiating-a-phate-estimator">3.3  Instantiating a PHATE estimator</h2>

<p>In practice, I usually start by running PHATE on a new dataset with default parameters, which is <code>n_components=2</code>, <code>knn=5</code>, and <code>decay=40</code>. Rarely, I will want to change these, but we will cover parameter tuning in a later section.</p>

<p>We can create a new <code>PHATE</code> object just like we did with the <code>PCA</code> class:</p>

<pre><code class="language-python">import phate
phate_op = phate.PHATE()
</code></pre>

<p>and now we can generate an embedding by calling <code>fit_transform</code> and plot the output using <code>scprep.plot.scatter2d</code>. Let&rsquo;s start with the T cell data from Datlinger et al. (2017).</p>

<pre><code class="language-python">Y = phate_op.fit_transform(data_sq)
</code></pre>

<p>Note: you can also pass in the PCA-reduced data from earlier, but PHATE can also do this for you at the cost of increased compute time.</p>

<p>You&rsquo;ll see some output that looks like this, telling you that PHATE is doing stuff:</p>

<pre><code>Calculating PHATE...
  Calculating graph and diffusion operator...
    Calculating PCA...
    Calculated PCA in 4.20 seconds.
    Calculating KNN search...
    Calculated KNN search in 1.23 seconds.
    Calculating affinities...
    Calculated affinities in 0.18 seconds.
  Calculated graph and diffusion operator in 6.20 seconds.
  Calculating landmark operator...
    Calculating SVD...
    Calculated SVD in 0.51 seconds.
    Calculating KMeans...
    Calculated KMeans in 14.68 seconds.
  Calculated landmark operator in 16.57 seconds.
  Calculating optimal t...
    Automatically selected t = 28
  Calculated optimal t in 0.80 seconds.
  Calculating diffusion potential...
  Calculated diffusion potential in 0.45 seconds.
  Calculating metric MDS...
  Calculated metric MDS in 33.45 seconds.
Calculated PHATE in 57.48 seconds.
</code></pre>

<p>Note, these times with setting <code>n_jobs=16</code>, you may find slower performance on a laptop or when using sparse input.</p>

<h2 id="3-4-plotting-phate-with-a-scatter-plot">3.4 Plotting PHATE with a scatter plot</h2>

<p>Now we can plot using <code>scatter2d</code>:</p>

<pre><code class="language-python">scprep.plot.scatter2d(Y, ticks=None, label_prefix='PHATE', figsize=(5,5))
</code></pre>

<p>And see the following plot:</p>




<figure class="img-md">

<img src="/img/how_to_single_cell/PHATE.tcell.nocolor.png" />


</figure>

<p>From this plot, we can tell that there are several kinds of cells in the data set, but it&rsquo;s hard to tell much other than that without starting to add some color. Let&rsquo;s plot the condition label for each cell, the library size, and expression of a mitochonrial gene.</p>

<pre><code class="language-python"># Plotting mitochondrial expression
fig, ax = plt.subplots(1, figsize=(3.4,3))
scprep.plot.scatter2d(Y, ax=ax, c=t_cell_data_ln['MT-CO1'],
                      title='MT-CO1', ticks=False, label_prefix='PHATE')
fig.tight_layout()

# Plotting libsize
fig, ax = plt.subplots(1, figsize=(3.6,3))
scprep.plot.scatter2d(Y, ax=ax, c=metadata['library_size'],
                      title='Library Size', ticks=False, label_prefix='PHATE')
fig.tight_layout()

# Plotting condition label
fig, ax = plt.subplots(1, figsize=(4.4,3))
condition_cdict = {'stimulated':'#1a3263',
                   'unstimulated':'#f5564e'}
scprep.plot.scatter2d(Y, ax=ax, c=metadata['condition'], cmap=condition_cdict,
                      title='Condition', ticks=False, label_prefix='PHATE',
                     legend_anchor=(1,1))
fig.tight_layout()
</code></pre>

<p>And we see the following plots:</p>




<figure class="img-lg">

<img src="/img/how_to_single_cell/PHATE.tcell.colors.png" />


</figure>

<p>Examining these plots, we see that there are no regions enriched for high mitochondrial RNA (which would indicate apoptotic cells). Although there are some regions of the plot with higher library size than others, these cells are fairly well distributed over the plot. If we say a branch of cells shooting off the plot with very high or low library sizes, then we might want to revisit the filtering thresholds established during preprocessing.</p>

<p>Finally, when we examine the the distribution of condition labels, we see that there is a good amount of overlap between the two conditions. We&rsquo;ll get to characterizing the differences between these two conditions in a later tutorial. For now, check out our method, MELD (Manifold Enhancement of Latent Dimensions) on <a href="https://doi.org/10.1101/532846" target="_blank">BioRxiv (doi: <code>10.1101/532846</code>)</a>.</p>

<h2 id="3-5-plotting-phate-as-a-kde-plot">3.5 Plotting PHATE as a KDE plot</h2>

<p>Scatter plots are a useful way to display data. They also have the appeal of showing you every single point in your dataset. <em>Or do they?</em></p>

<p>Scatter plots have one drawback, which is that data points that are overlapping in the plot get drawn on top of each other. This makes it difficult to identify where most of the data density lies. You could color your plot by a density estimate, as is commonly done in FACS analysis. However, I find that in these plots, the eye is still drawn to outliers.</p>

<p>Instead, I think it&rsquo;s useful to look at a Kernel Density Estimate (KDE) plot. These plots use kernel functions (just like in the graph building process) to estimate data density in one or more dimensions.</p>

<p>In python you can make a KDE plot using the <a href="https://seaborn.pydata.org/" target="_blank"><code>seaborn</code></a> package. <code>seaborn</code> works seamlessly with <code>numpy</code> and <code>pandas</code> and uses <code>matplotlib</code> as a backend. The function for drawing a KDE plot is called, unsurprisingly <code>kdeplot</code>.</p>

<p>Let&rsquo;s generate a KDE plot of the T cell data to get an idea of how many clusters of cells are in the data.</p>

<pre><code class="language-python">import seaborn as sns

fig, ax = plt.subplots(1, figsize=(4,4))
sns.kdeplot(Y[:,0], Y[:,1], n_levels=100, shade=True, cmap='inferno', zorder=0, ax=ax)

ax.set_xticks([])
ax.set_yticks([])

ax.set_xlabel('PHATE 1', fontsize=18)
ax.set_ylabel('PHATE 2', fontsize=18)

ax.set_title('KDE - T cells', fontsize=20)

fig.tight_layout()
</code></pre>




<figure class="img-md">

<img src="/img/how_to_single_cell/PHATE.KDE.tcell.png" />


</figure>

<p>Here, it&rsquo;s much easier to see that there are around 5 regions of density in the data. Let&rsquo;s cluster them and figure out what they are!</p>

<h2 id="3-6-clustering-data-using-phate">3.6 Clustering data using PHATE</h2>

<p>A common mode of analysis for scRNA-seq is to identify clusters of cells and characterize the transcriptional diversity of these subpopulations. There are many clustering algorithms out there, and many of them have implementations in <a href="https://scikit-learn.org/stable/modules/clustering.html#clustering" target="_blank"><code>sklearn.cluster</code></a>. One of our favorite clustering algorithms is <a href="https://link.springer.com/article/10.1007/s11222-007-9033-z" target="_blank">Spectral Clustering</a>. In the PHATE package, we have developed an adaptation of this method using the PHATE diffusion potential instead of the eigenvectors of the normalized Laplacian.</p>

<p>You can call this method with <code>phate.cluster.kmeans</code>, and use these labels to color the PHATE plot. We&rsquo;ll set <code>k=5</code> based on our examination of the KDE plot.</p>

<pre><code class="language-python">clusters = phate.cluster.kmeans(phate_op, k=5)

scprep.plot.scatter2d(Y, c=clusters, cmap=sns.husl_palette(5), s=1,
                      figsize=(4.3,4), ticks=None, label_prefix='PHATE',
                     legend_anchor=(1,1), fontsize=12, title='PHATE clusters')

fig.tight_layout()
</code></pre>




<figure class="img-md">

<img src="/img/how_to_single_cell/PHATE.tcell.clusters.png" />


</figure>

<p>Here, we can see that the clusters are localized on the PHATE plot, and now we can begin to characterize them. Note, I won&rsquo;t claim that this is the best method for identifying clusters. In general, cluster assignments should be seen as &ldquo;best guess&rdquo; partitions of the data. For every clustering algorithm, you can change some parameters and increase or decrease the number of clusters identified. The best way to make sure you have &ldquo;valid&rdquo; clusters is to inspect each group individually, and make sure you&rsquo;re happy with the resolution of the results.</p>

          </div>

          

        </div>

        <div class="body-footer">
          Last updated on Jan 1, 2019
        </div>

      </article>

      <footer class="site-footer">
  

  <p class="powered-by">
    &copy; 2019 Krishnaswamy Lab &middot; 

  </p>
</footer>


    </main>
  </div>
</div>

    

    
    
    
    <script src="/js/mathjax-config.js"></script>
    

    
    
    
      <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha512-+NqPlbbtM1QqiK8ZAo4Yrj2c4lNQoGv8P79DPtKzj++l5jnN39rHA/xsqn8zE9l0uSoxaCdrOgFs6yjyfbBxSg==" crossorigin="anonymous"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.imagesloaded/4.1.3/imagesloaded.pkgd.min.js" integrity="sha512-umsR78NN0D23AzgoZ11K7raBD+R6hqKojyBZs1w8WvYlsI+QuKRGBx3LFCwhatzBunCjDuJpDHwxD13sLMbpRA==" crossorigin="anonymous"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/4.1.3/js/bootstrap.min.js" integrity="sha256-VsEqElsCHSGmnmHXGQzvoWjWwoznFSZc6hs7ARLRacQ=" crossorigin="anonymous"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.isotope/3.0.4/isotope.pkgd.min.js" integrity="sha512-VDBOIlDbuC4VWxGJNmuFRQ0Li0SKkDpmGyuhAG5LTDLd/dJ/S0WMVxriR2Y+CyPL5gzjpN4f/6iqWVBJlht0tQ==" crossorigin="anonymous"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.2.5/jquery.fancybox.min.js" integrity="sha256-X5PoE3KU5l+JcX+w09p/wHl9AzK333C4hJ2I9S5mD4M=" crossorigin="anonymous"></script>

      
        
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js" integrity="sha256-/BfiIkHlHoVihZdc6TFuj7MmJ0TWcWsMXkeDFwhi0zw=" crossorigin="anonymous"></script>
        
      

      
      
      <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.4/MathJax.js?config=TeX-AMS_CHTML-full" integrity="sha256-GhM+5JHb6QUzOQPXSJLEWP7R73CbkisjzK5Eyij4U9w=" crossorigin="anonymous" async></script>
      
    

    
    

    
    
    

    
    
    <script>hljs.initHighlightingOnLoad();</script>
    

    
    
    <script>
      const search_index_filename = "/index.json";
      const i18n = {
        'placeholder': "Search...",
        'results': "results found",
        'no_results': "No results found"
      };
      const content_type = {
        'post': "Posts",
        'project': "Projects",
        'publication' : "Publications",
        'talk' : "Talks"
        };
    </script>
    

    
    
    <script src="https://cdnjs.cloudflare.com/ajax/libs/anchor-js/4.1.1/anchor.min.js" integrity="sha256-pB/deHc9CGfFpJRjC43imB29Rse8tak+5eXqntO94ck=" crossorigin="anonymous"></script>
    <script>
      anchors.add();
    </script>
    

    
    
    <script id="search-hit-fuse-template" type="text/x-template">
      <div class="search-hit" id="summary-{{key}}">
      <div class="search-hit-content">
        <div class="search-hit-name">
          <a href="{{relpermalink}}">{{title}}</a>
          <div class="article-metadata search-hit-type">{{type}}</div>
          <p class="search-hit-description">{{snippet}}</p>
        </div>
      </div>
      </div>
    </script>
    

    
    
    <script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/3.2.1/fuse.min.js" integrity="sha256-VzgmKYmhsGNNN4Ph1kMW+BjoYJM2jV5i4IlFoeZA9XI=" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script>
    

    
    

    
    
    
    
    
    
    
    
    <script src="/js/academic.min.70f0041f5a24c6a675ac218c98d7ef71.js"></script>

    

  </body>
</html>


